perma_web/api2/serializers.py

```diff
@@ -0,0 +1,242 @@
+        return url
+
+    def validate(self, data):
+        user = self.context['request'].user
+        errors = {}
+
+        # handle private_reason:
+        if not user.is_staff:
+            # only staff can manually set private_reason
+            data.pop('private_reason', None)
+        if self.instance:
+            # if updating privacy, make sure user is allowed to change private status
+            if not user.is_staff and 'is_private' in data and self.instance.is_private != bool(data['is_private']):
+                if self.instance.private_reason and self.instance.private_reason != 'user':
+                    errors['is_private'] = 'Cannot change link privacy.'
+                else:
+                    data['private_reason'] = 'user' if data['is_private'] else None
+        else:
+            # for new links, set private_reason based on is_private
+            data['private_reason'] = 'user' if data.get('is_private') else None
+
+        # check submitted URL for new link
+        if not self.instance:
+            if not data.get('submitted_url'):
+                errors['url'] = "URL cannot be empty."
+            else:
+                try:
+                    validate = URLValidator()
+                    temp_link = Link(submitted_url=data['submitted_url'])
+                    validate(temp_link.safe_url)
+
+                    # Don't force URL resolution validation if a file is provided
+                    if not data.get('file'):
+                        if not temp_link.ip:
+                            errors['url'] = "Couldn't resolve domain."
+                        elif not ip_in_allowed_ip_range(temp_link.ip):
+                            errors['url'] = "Not a valid IP."
+                        elif not temp_link.headers:
+                            errors['url'] = "Couldn't load URL."
+                        else:
+                            # preemptively reject URLs that report a size over settings.MAX_ARCHIVE_FILE_SIZE
+                            try:
+                                if int(temp_link.headers.get('content-length', 0)) > settings.MAX_ARCHIVE_FILE_SIZE:
+                                    errors['url'] = "Target page is too large (max size %sMB)." % (settings.MAX_ARCHIVE_FILE_SIZE / 1024 / 1024)
+                            except ValueError:
+                                # content-length header wasn't an integer. Carry on.
+                                pass
+                except UnicodeError:
+                    # see https://github.com/harvard-lil/perma/issues/1841
+                    errors['url'] = "Unicode error while processing URL."
+                except DjangoValidationError:
+                    errors['url'] = "Not a valid URL."
+                except TooManyRedirects:
+                    errors['url'] = "URL caused a redirect loop."
+
+        # check uploaded file
+        uploaded_file = self.context['request'].data.get('file')
```

Comment:
No, because 'file' isn't actually a field on the model -- by this point DRF has filtered down data to just known fields. I'll add a comment to explain why we access that in a different way.